temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+            # with the obvious exception of our context (if they want more, they
+            # can set the initializer on the executor).
+            ctx = temporalio.activity._Context.current()
+            info = ctx.info()
+
+            # Heartbeat calls internally use a data converter which is async so
+            # they need to be called on the event loop
+            loop = asyncio.get_running_loop()
+            orig_heartbeat = ctx.heartbeat
+
+            def thread_safe_heartbeat(*details: Any) -> None:
+                temporalio.activity._Context.set(ctx)
+                loop.call_soon_threadsafe(orig_heartbeat, *details)
+
+            ctx.heartbeat = thread_safe_heartbeat
+
+            # For heartbeats, we use the existing heartbeat callable for thread
+            # pool executors or a multiprocessing queue for others
+            heartbeat: Union[Callable[..., None], SharedHeartbeatSender] = ctx.heartbeat
+            shared_manager: Optional[SharedStateManager] = None
+            if not isinstance(input.executor, concurrent.futures.ThreadPoolExecutor):
+                # Should always be present in worker, pre-checked on init
+                shared_manager = input._worker._config["shared_state_manager"]
+                assert shared_manager
+                heartbeat = shared_manager.register_heartbeater(
+                    info.task_token, ctx.heartbeat
+                )
+
+            try:
+                return await asyncio.get_running_loop().run_in_executor(
+                    input.executor,
+                    _execute_sync_activity,
+                    info,
+                    heartbeat,
+                    # Only thread event, this may cross a process boundary
+                    input._cancelled_event.thread_event,
+                    input._worker_shutdown_event.thread_event,
+                    input.fn,
+                    *input.args,
+                )
+            finally:
+                if shared_manager:
+                    shared_manager.unregister_heartbeater(info.task_token)
+
+        # Otherwise for async activity, just run
+        return await input.fn(*input.args)
+
+
+# This has to be defined at the top-level to be picklable for process executors
+def _execute_sync_activity(
+    info: temporalio.activity.Info,
+    heartbeat: Union[Callable[..., None], SharedHeartbeatSender],
+    cancelled_event: threading.Event,
+    worker_shutdown_event: threading.Event,
+    fn: Callable[..., Any],
+    *args: Any,
+) -> Any:
+    heartbeat_fn: Callable[..., None]
+    if isinstance(heartbeat, SharedHeartbeatSender):
+        # To make mypy happy
+        heartbeat_sender = heartbeat
+        heartbeat_fn = lambda *details: heartbeat_sender.send_heartbeat(
+            info.task_token, *details
+        )
+    else:
+        heartbeat_fn = heartbeat
+    temporalio.activity._Context.set(
+        temporalio.activity._Context(
+            info=lambda: info,
```

Comment:
Hrmm, this removes the ability for an interceptor to override this for non-async activities, but do I want to make a whole threadsafe-and-picklable info call?